File: python/python_asyncflow/logger.py
"""Logging utilities for AsyncFlow"""

import logging
from typing import Optional


class Logger:
    """Simple logger wrapper for AsyncFlow components"""
    
    def __init__(self, prefix: str = "AsyncFlow", verbosity: int = 3):
        self.prefix = prefix
        self.verbosity = verbosity
        
        # Setup Python logging
        logging.basicConfig(
            level=logging.DEBUG,
            format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
        )
        self._logger = logging.getLogger(prefix)
    
    def error(self, message: str) -> None:
        """Log error message"""
        self._logger.error(f"[{self.prefix}] {message}")
    
    def warn(self, message: str) -> None:
        """Log warning message"""
        self._logger.warning(f"[{self.prefix}] {message}")
    
    def info(self, message: str) -> None:
        """Log info message"""
        self._logger.info(f"[{self.prefix}] {message}")
    
    def debug(self, message: str) -> None:
        """Log debug message"""
        self._logger.debug(f"[{self.prefix}] {message}")
    
    def verbose(self, level: int, message: str) -> None:
        """Log message at specific verbosity level"""
        if level <= self.verbosity:
            if level == 0:
                self.error(message)
            elif level == 1:
                self.warn(message)
            elif level == 2:
                self.info(message)
            else:
                self.debug(message)

File: src/logger.rs
use tracing::{debug, error, info, trace, warn};

#[derive(Debug, Clone, Copy)]
pub enum LogLevel {
    Error = 0,
    Warn = 1,
    Info = 2,
    Debug = 3,
    Trace = 4,
}

#[derive(Debug, Clone)]
pub struct Logger {
    pub verbosity: u8,
    pub prefix: String,
}

impl Logger {
    pub fn new(prefix: impl Into<String>) -> Self {
        Self {
            verbosity: 3,
            prefix: prefix.into(),
        }
    }

    pub fn error(&self, message: impl AsRef<str>) {
        error!("[{}] {}", self.prefix, message.as_ref());
    }

    pub fn warn(&self, message: impl AsRef<str>) {
        warn!("[{}] {}", self.prefix, message.as_ref());
    }

    pub fn info(&self, message: impl AsRef<str>) {
        info!("[{}] {}", self.prefix, message.as_ref());
    }

    pub fn debug(&self, message: impl AsRef<str>) {
        debug!("[{}] {}", self.prefix, message.as_ref());
    }

    pub fn trace(&self, message: impl AsRef<str>) {
        trace!("[{}] {}", self.prefix, message.as_ref());
    }

    pub fn verbose(&self, level: u8, message: impl AsRef<str>) {
        if level <= self.verbosity {
            match level {
                0 => self.error(message),
                1 => self.warn(message),
                2 => self.info(message),
                3 => self.debug(message),
                _ => self.trace(message),
            }
        }
    }
}

impl Default for Logger {
    fn default() -> Self {
        Self::new("AsyncFlow")
    }
}